tmpfiles: add factory reset type modifier

bluca opened the pull request:
If '_' is prefixed to an entry type, consider it only if --factory-reset
is passed to systemd-tmpfiles. Useful to remove files/directories only
when a system is being reset.

yuwata reviewed (approved):
LGTM.

yuwata commented:
@mrc0mmand Could you restart the failed CentOS CI job?

DaanDeMeyer commented:
What about '@'? I don't think we use '_' anywhere yet, there's some precedence for '@' at least.

poettering commented:
This removes the TODO list item entirely, even though it suggests two distinct things: conditionalize on first boot, and conditionalize on factory reset. The item regarding the first boot conditionalization should stay, given this only addresses the factory reset

bluca commented on src/tmpfiles/tmpfiles.c:
```diff
@@ -3743,6 +3754,10 @@ static int parse_argv(int argc, char *argv[]) {
+                case ARG_FACTORY_RESET:
+                        arg_factory_reset = true;
+                        break;
+
```
I can change the option to take a boolean - the rest sounds like it can be done separately? I want to use this to let packages define things that need to be done on 'purge' (which is different from 'remove')

mbiebl replied:
I assume you are referring to the distinction between remove and purge in Debian.
While I like the idea to have a mechanism to cleanup resources when a package is uninstalled in a declarative way, I'm unsure how this could be done via tmpfiles. Those files are removed during the "uninstall" phase, so they might no longer be available during "purge".

bluca replied:
That's our problem to figure out - we can do it in the prerm for example

bluca replied:
We could copy it out to /var/ somewhere in the prerm, and call it in the postrm+purge, or something like that. We don't have to solve it here - the important thing here is to have the functionality available

poettering replied:
> I can change the option to take a boolean - the rest sounds like it can be done separately? I want to use this to let packages define things that need to be done on 'purge' (which is different from 'remove')

I don't understand what this means? I sense this is an entirely different usecase for factory reset than the systemd-repart thing?

can you elaborate the usecase, I really don't grok this?

poettering replied:
i don't get it? why involve tmpfiles in package management? doesn't dpkg track file ownership?

bluca replied:
it's for things that are _created_ with tmpfiles.d - ie, you ship a tmpfiles to create stuff in /var/, and you add a `R@` line so that it gets removed on purge, but not when the package is simply removed, or `tmpfiles --remove` is ran for other reasons

poettering replied:
but why doesn't dpkg track that? on rpm you gave "ghost" entries in RPM ownership definition. dpkg doesn't have that?

bluca replied:
no, there's no such thing

poettering replied:
that sounds like something to fix in dpkg though, no?

bluca replied:
yeah that's not, er, likely to happen

bluca commented:
> What about '@'? I don't think we use '_' anywhere yet, there's some precedence for '@' at least.

changed to '@'

bluca commented:
@poettering I've changed the flag to boolean, and I've added to the TODO list the rest of the features you mentioned - I don't think it needs to be all done at the same time

yuwata commented on src/tmpfiles/tmpfiles.c:
```diff
@@ -198,6 +198,7 @@ static bool arg_user = false;
 static OperationMask arg_operation = 0;
 static bool arg_boot = false;
 static PagerFlags arg_pager_flags = 0;
+static int arg_factory_reset = -1;
```
Why this is tristate, instead of boolean?? Only for logging??

bluca replied:
See top of this comment: https://github.com/systemd/systemd/pull/24983#discussion_r994341848

yuwata replied:
So, IIUC, this is for future extension, reading settings from EFI variable, right?

bluca replied:
I guess so, but this way the setting will stay compatible

yuwata replied:
I think it is better to use `bool` now, and change it later when such feature is implemented. But OK. Not a strong opinion.

poettering commented on TODO:
```diff
@@ -181,8 +181,17 @@ Features:
+* systemd-tmpfiles: add concept for conditionalizing lines on first boot.
+
+* systemd-tmpfiles: check the kernel cmdline and EFI variable for factory reset
+  like systemd-repart does, and use it even when --factory-reset=no
```
hmm, no, if --factory-reset=no is set, we should *not* check the efi var and kernel cmdline.

if --factory-reset= is explicitly specified it should always take precedence. Only if it is not specified we should check efi var/kernel cmdline

poettering commented on TODO:
```diff
@@ -181,8 +181,17 @@ Features:
+  like systemd-repart does, and use it even when --factory-reset=no
+
+* add systemd-factory-reset-complete.service that removes the factory reset EFI
+  variable and resets the GPT disk UUID
```
also, please say that the logic for this should be removed from systemd-repart

poettering commented on man/tmpfiles.d.xml:
```diff
@@ -482,6 +482,15 @@ r! /tmp/.X[0-9]*-lock</programlisting>
 
+      <para>If the at character (<literal>@</literal>) is used, this line is only safe to execute during a factory
+      reset, and can break a running system. <command>systemd-tmpfiles</command> will take lines with
+      an at character only into consideration, if the <option>--factory-reset</option> option is given.</para>
```
`--factory-reset=` *with* a trailing equal sign, please

poettering commented on man/tmpfiles.d.xml:
```diff
@@ -482,6 +482,15 @@ r! /tmp/.X[0-9]*-lock</programlisting>
       running system, and will only be executed with
       <option>--boot</option>.</para>
 
+      <para>If the at character (<literal>@</literal>) is used, this line is only safe to execute during a factory
```
it's not about safety really, it's about the explicit equest to do this.

poettering commented:
hmm, I think there is some disparity between the "factory reset" concept you want to use this for and the one otherwise assumed. Can you elaborate what you are trying to do?

I also wonder if the whole approach is actually the right approach. how would people actually use this?

let's say people want to use this to reset /etc/motd to some "hey, welcome" text whenever the system is factory reset. So, they want two things: probably one line that ensures that if the file is missing it is written with some contents, and another line that declares it shall be removed in case of factory reset. But the `@` flag wouldn't be able to express that, you couldn't conditionalize the remve line, because it would then still conflict with the write line...

So maybe: instead of conditionalizing lines just duplicate the whole tmpfiles.d/ pick-up, i.e. introduce /usr/lib/tmpfiles-factory-reset.d/ (and similar under (/etc/, /run/, …) or so. And the --factory-reset= switch then simply controls whether to look into that, or into the regular /usr/lib/tmpfiles.d/.

And then wrap an invocation for that in a new systemd-tmpfiles-factory-reset.service instance, tht is properly conditioned, with COnditionKernelCommandLine= and ConditionFileExists= with a check in the efivarfs

bluca commented:
> hmm, I think there is some disparity between the "factory reset" concept you want to use this for and the one otherwise assumed. Can you elaborate what you are trying to do?
> 
> I also wonder if the whole approach is actually the right approach. how would people actually use this?
> 
> let's say people want to use this to reset /etc/motd to some "hey, welcome" text whenever the system is factory reset. So, they want two things: probably one line that ensures that if the file is missing it is written with some contents, and another line that declares it shall be removed in case of factory reset. But the `@` flag wouldn't be able to express that, you couldn't conditionalize the remve line, because it would then still conflict with the write line...
> 
> So maybe: instead of conditionalizing lines just duplicate the whole tmpfiles.d/ pick-up, i.e. introduce /usr/lib/tmpfiles-factory-reset.d/ (and similar under (/etc/, /run/, …) or so. And the --factory-reset= switch then simply controls whether to look into that, or into the regular /usr/lib/tmpfiles.d/.
> 
> And then wrap an invocation for that in a new systemd-tmpfiles-factory-reset.service instance, tht is properly conditioned, with COnditionKernelCommandLine= and ConditionFileExists= with a check in the efivarfs

I don't quite like the idea of a new location that has to be supported by everything, and also don't fully understand the use case - why wouldn't this work?

```
R@ /etc/motd
f /etc/motd - - - - welcome
```

this seems to work as you intend?

poettering commented:
> ```
> 
> ```
> 
> 
>     
>   
> 
> this seems to work as you intend?

the lines are conflicing, because for the same path

poettering commented:
I am not convinced tmpfiles is really the tool of choice to fake file ownership in dpkg i must say.

poettering commented:
I mean, the factory reset logic so far has a very clear meaning: the whole system is reset back to zero config during bootup. but if you start to use it for removing packaging artifacts, then is this *really* the same stuff you want to remove during a factory reset boot? I really doubt that.

i.e. let's say i build an image with "gdm", and want to ensure that certain settings in gdm.conf match my expectations when a factory reset is done. i.e. i want to reset gdm.conf to some specific defaults. But this is quite a different operation from what should happen during dpkg package removal, where you don't actually want to reset, but just remove. These are hence two distinct things, and they shouldn't be bound to the same concept...

bluca commented:
> the lines are conflicing, because for the same path

I'm not sure what you mean by conflicting - it just works?

```
$ echo bar > /tmp/foo
$ echo -e "R@ /tmp/foo\nf /tmp/foo - - - - baz\n" | buildx86/systemd-tmpfiles --create --remove -
$ cat /tmp/foo
bar
$ echo -e "R@ /tmp/foo\nf /tmp/foo - - - - baz\n" | buildx86/systemd-tmpfiles --create --remove --factory-reset=yes -
$ cat /tmp/foo
baz
```

poettering commented:
hmm, that sounds like a bug. the two lines should conflict if one says "this should exist" and the other says "no it shouldnt"

bluca commented:
> hmm, that sounds like a bug. the two lines should conflict if one says "this should exist" and the other says "no it shouldnt"

too late, it's a feature now!

bluca commented:
> I mean, the factory reset logic so far has a very clear meaning: the whole system is reset back to zero config during bootup. but if you start to use it for removing packaging artifacts, then is this _really_ the same stuff you want to remove during a factory reset boot? I really doubt that.
> 
> i.e. let's say i build an image with "gdm", and want to ensure that certain settings in gdm.conf match my expectations when a factory reset is done. i.e. i want to reset gdm.conf to some specific defaults. But this is quite a different operation from what should happen during dpkg package removal, where you don't actually want to reset, but just remove. These are hence two distinct things, and they shouldn't be bound to the same concept...

That still works fine, as we'll call tmpfiles --remove, so nothing will be created, and only removals will be done

bluca commented:
ok, I've added the factory-reset-complete target and services as requested

mrc0mmand commented:
Interesting, one of the failed TEST-29-PORTABLE caught this:

```
[   49.462982] systemd-portabled[539]: ==539==ERROR: LeakSanitizer: detected memory leaks
[   49.462982] systemd-portabled[539]: Direct leak of 160 byte(s) in 1 object(s) allocated from:
[   49.477805] systemd-portabled[539]:     #0 0x7f841a30a70a  (/usr/lib/clang/14.0.6/lib/linux/libclang_rt.asan-x86_64.so+0x10a70a) (BuildId: 53dba8957703192f5863771c03d512bbb9f0b252)
[   49.477805] systemd-portabled[539]:     #1 0x7f84198365d8 in greedy_realloc /systemd-meson-build/../build/src/basic/alloc-util.c:70:13
[   49.477805] systemd-portabled[539]:     #2 0x7f84199a3dde in bus_match_parse /systemd-meson-build/../build/src/libsystemd/sd-bus/bus-match.c:801:22
[   49.477805] systemd-portabled[539]:     #3 0x7f8419a3d376 in bus_add_match_full /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3495:13
[   49.477805] systemd-portabled[539]:     #4 0x7f8419a3dfc1 in sd_bus_add_match_async /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3581:16
[   49.477805] systemd-portabled[539]:     #5 0x7f841958fee0 in bus_async_unregister_and_exit /systemd-meson-build/../build/src/shared/bus-util.c:82:13
[   49.477805] systemd-portabled[539]:     #6 0x7f841959042a in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:130:29
[   49.477805] systemd-portabled[539]:     #7 0x55e33e288661 in manager_run /systemd-meson-build/../build/src/portable/portabled.c:119:16
[   49.477805] systemd-portabled[539]:     #8 0x55e33e288661 in run /systemd-meson-build/../build/src/portable/portabled.c:161:13
[   49.477805] systemd-portabled[539]:     #9 0x55e33e288661 in main /systemd-meson-build/../build/src/portable/portabled.c:168:1
[   49.477805] systemd-portabled[539]:     #10 0x7f8418e3c28f  (/usr/lib/libc.so.6+0x2328f) (BuildId: 1e94beb079e278ac4f2c8bce1f53091548ea1584)
[   49.477805] systemd-portabled[539]: Indirect leak of 384 byte(s) in 6 object(s) allocated from:
[   49.477805] systemd-portabled[539]:     #0 0x7f841a30a70a  (/usr/lib/clang/14.0.6/lib/linux/libclang_rt.asan-x86_64.so+0x10a70a) (BuildId: 53dba8957703192f5863771c03d512bbb9f0b252)
[   49.477805] systemd-portabled[539]:     #1 0x7f84198365d8 in greedy_realloc /systemd-meson-build/../build/src/basic/alloc-util.c:70:13
[   49.477805] systemd-portabled[539]:     #2 0x7f84199a3aa7 in bus_match_parse /systemd-meson-build/../build/src/libsystemd/sd-bus/bus-match.c:775:30
[   49.477805] systemd-portabled[539]:     #3 0x7f8419a3d376 in bus_add_match_full /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3495:13
[   49.477805] systemd-portabled[539]:     #4 0x7f8419a3dfc1 in sd_bus_add_match_async /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3581:16
[   49.477805] systemd-portabled[539]:     #5 0x7f841958fee0 in bus_async_unregister_and_exit /systemd-meson-build/../build/src/shared/bus-util.c:82:13
[   49.613309] systemd-portabled[539]:     #6 0x7f841959042a in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:130:29
[   49.613309] systemd-portabled[539]:     #7 0x55e33e288661 in manager_run /systemd-meson-build/../build/src/portable/portabled.c:119:16
[   49.613309] systemd-portabled[539]:     #8 0x55e33e288661 in run /systemd-meson-build/../build/src/portable/portabled.c:161:13
[   49.613309] systemd-portabled[539]:     #9 0x55e33e288661 in main /systemd-meson-build/../build/src/portable/portabled.c:168:1
[   49.613309] systemd-portabled[539]:     #10 0x7f8418e3c28f  (/usr/lib/libc.so.6+0x2328f) (BuildId: 1e94beb079e278ac4f2c8bce1f53091548ea1584)
[   49.613309] systemd-portabled[539]: Indirect leak of 1 byte(s) in 1 object(s) allocated from:
[   49.613309] systemd-portabled[539]:     #0 0x7f841a2d25a2 in strdup (/usr/lib/clang/14.0.6/lib/linux/libclang_rt.asan-x86_64.so+0xd25a2) (BuildId: 53dba8957703192f5863771c03d512bbb9f0b252)
[   49.613309] systemd-portabled[539]:     #1 0x7f84199a3d35 in bus_match_parse /systemd-meson-build/../build/src/libsystemd/sd-bus/bus-match.c:785:33
[   49.613309] systemd-portabled[539]:     #2 0x7f8419a3d376 in bus_add_match_full /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3495:13
[   49.613309] systemd-portabled[539]:     #3 0x7f8419a3dfc1 in sd_bus_add_match_async /systemd-meson-build/../build/src/libsystemd/sd-bus/sd-bus.c:3581:16
[   49.613309] systemd-portabled[539]:     #4 0x7f841958fee0 in bus_async_unregister_and_exit /systemd-meson-build/../build/src/shared/bus-util.c:82:13
[   49.613309] systemd-portabled[539]:     #5 0x7f841959042a in bus_event_loop_with_idle /systemd-meson-build/../build/src/shared/bus-util.c:130:29
[   49.613309] systemd-portabled[539]:     #6 0x55e33e288661 in manager_run /systemd-meson-build/../build/src/portable/portabled.c:119:16
[   49.613309] systemd-portabled[539]:     #7 0x55e33e288661 in run /systemd-meson-build/../build/src/portable/portabled.c:161:13
[   49.613309] systemd-portabled[539]:     #8 0x55e33e288661 in main /systemd-meson-build/../build/src/portable/portabled.c:168:1
[   49.613309] systemd-portabled[539]:     #9 0x7f8418e3c28f  (/usr/lib/libc.so.6+0x2328f) (BuildId: 1e94beb079e278ac4f2c8bce1f53091548ea1584)
[   49.613309] systemd-portabled[539]: SUMMARY: AddressSanitizer: 545 byte(s) leaked in 8 allocation(s).
```

Not sure if it's caused by this PR, but it's definitely worth looking into.

bluca commented:
> Not sure if it's caused by this PR, but it's definitely worth looking into.

Weird but definitely not from this PR, not touching either dbus or one of the services here

mrc0mmand commented:
> > Not sure if it's caused by this PR, but it's definitely worth looking into.
> 
> Weird but definitely not from this PR, not touching either dbus or one of the services here

Filed as https://github.com/systemd/systemd/issues/25340 so it won't get forgotten.

DaanDeMeyer commented:
I figured the semantics behind factory reset would be nuking /etc completely so that we start from a known state. If you start removing individual files from a modified system, you'll never be sure that you're back in a pristine state after factory reset. So shouldn't factory reset just be limited to completely nuking /etc and other directories containing state instead of adding functionality to incrementally remove files from the system?

bluca commented:
> I figured the semantics behind factory reset would be nuking /etc completely so that we start from a known state. If you start removing individual files from a modified system, you'll never be sure that you're back in a pristine state after factory reset. So shouldn't factory reset just be limited to completely nuking /etc and other directories containing state instead of adding functionality to incrementally remove files from the system?

Well, it depends, for a pure image-based system, it's possible, but we are still far away from being realistically able to handle a system with an empty /etc. Once that becomes possible, this won't stop that from working.

For package-based ones, reset works at the package level too (ie: `apt purge foo`), so we need to have functionality that matches the install step with the purge one, and there isn't one yet. This will remain useful regardless of the above. This is my main use case.

keszybz commented on test/units/testsuite-22.15.sh:
```diff
@@ -0,0 +1,34 @@
+
+rm -rf /tmp/somedir
+mkdir -p /tmp/somedir
+echo bar > /tmp/somedir/somefile
```
Please, no space after redirection op.

keszybz reviewed (commented):
Looks OK. I have to admit that I'm not 100% that we want this at all — it still seems to me that the whole factory reset logic in systemd is something untested and incomplete. But as far as the implementation here goes, it looks solid: it'll do what it's supposed to do. So I think it's fine to merge this — it may turn out to good enough for some use cases. If not, no big harm.

bluca commented:
> I'm leaving `please-review` and I won't merge this myself, so somebody else has a chance to bless it.

ok, I'll leave it up too

poettering commented:
hmm, at fosdem we discussed changing the logic of this to a concept of "ownership", right? i.e. that we add a new switch "--purge" to systemd-tmpfiles that removes all files a specific tmpfiles file creates.

bluca commented:
> hmm, at fosdem we discussed changing the logic of this to a concept of "ownership", right? i.e. that we add a new switch "--purge" to systemd-tmpfiles that removes all files a specific tmpfiles file creates.

Yes, hence the needs-work label, working on it

bluca commented:
--> https://github.com/systemd/systemd/pull/30414
